File: DDModelLibrary/src/main/java/com/ddmodel/Block.java
package com.ddmodel;

import com.ddmodel.ability.Ability;
import com.ddmodel.ability.AbilityType;
import com.ddmodel.feature.Feature;
import com.ddmodel.stat.Stat;
import com.ddmodel.stat.StatType;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Block {
    private String name;
    private TypedSet<Ability> abilityScores = new TypedSet<>();
    private TypedSet<Stat> stats = new TypedSet<>();
    private TypedSet<Feature> features = new TypedSet<>();
    private TypedSet<Action> actions = new TypedSet<>();

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<Ability> getAbilityScores() {
        return new ArrayList<>(abilityScores);
    }

    public Optional<Ability> getAbility(AbilityType type) {
        return abilityScores.getValueMatching(ability -> ability.getAbilityType().equals(type));
    }

    public void setAbilityScores(TypedSet<Ability> abilityScores) {
        this.abilityScores = abilityScores;
    }

    public TypedSet<Stat> getStats() {
        return stats;
    }

    public void setStats(TypedSet<Stat> stats) {
        this.stats = stats;
    }

    public TypedSet<Feature> getFeatures() {
        return features;
    }

    public void setFeatures(TypedSet<Feature> features) {
        this.features = features;
    }

    public TypedSet<Action> getActions() {
        return actions;
    }

    public void setActions(TypedSet<Action> actions) {
        this.actions = actions;
    }

    public Optional<Stat> getStat(StatType type) {
        return stats.getValueMatching(stat -> stat.getType().equals(type));
    }

    public void putAbility(Ability ability) {
        abilityScores.put(ability);
    }

    public void putStat(Stat stat) {
        stats.put(stat);
    }

    public void putFeature(Feature feature) {
        features.put(feature);
    }

    public void putAction(Action action) {
        actions.put(action);
    }

    @Override
    public String toString() {
        return "Name : " + name + '\n' +
                "Abilities : " + abilityScores + '\n' +
                "Statistics : " + stats + '\n' +
                "Actions : " + actions + '\n' +
                "Features : " + features;
    }
}

File: engine/src/main/scala/com/pastorm/model/Block.scala
package com.pastorm.model

import com.pastorm.model.dice.Die

import scala.scalajs.js.annotation.{JSExport, JSExportAll, JSExportTopLevel}
import upickle.default._
import upickle.default.{macroRW, ReadWriter => RW}

@JSExportAll
sealed trait StatValue {
  def formulae: String

  def meanValue: Int

  def instantiateValue: Int
}
object StatValue{
  implicit val rw: RW[StatValue] = macroRW
}

@JSExportTopLevel("ConstValue")
@JSExportAll
case class ConstValue(formulae: String, meanValue: Int) extends StatValue {
  override def instantiateValue: Int = meanValue
}
object ConstValue{
  implicit val rw: RW[ConstValue] = macroRW
}

@JSExportTopLevel("DiceValue")
@JSExportAll
case class DiceValue(number: Int, sides: Int, modifier: Int) extends StatValue {
  override def instantiateValue: Int = {
    val die = new Die(sides)
    var acc = modifier
    for (_ <- 0 until number) {
      acc += die.roll
    }
    acc
  }

  override def meanValue: Int = {
    val facesUp : Double = sides + 1
    val div : Double = facesUp / 2
    val res : Double = number * div + modifier
    res.toInt
  }

  override def formulae: String = {
    val modifierFormat = s"${if (modifier > 0) "+" else "-"}$modifier"
    s"${number}d$sides${if (modifier != 0) modifierFormat else ""}"
  }
}
object DiceValue{
  implicit val rw: RW[DiceValue] = macroRW
}

case class Ability(abilityType: String, score: Int, modifier: Int)
object Ability{
  implicit val rw: RW[Ability] = macroRW
}
case class Stat(statType: String, statValue: StatValue)
object Stat{
  implicit val rw: RW[Stat] = macroRW
}
case class Feature(name: String, description: String)
object Feature{
  implicit val rw: RW[Feature] = macroRW
}
@JSExportTopLevel("Action")
@JSExportAll
case class Action(name: String,
                  toHit: String,
                  reach: String,
                  range: String,
                  hit: String,
                  description: String)
object Action{
  implicit val rw: RW[Action] = macroRW
}

case class Block(name: String,
                 abilityScores: List[Ability],
                 stats: List[Stat],
                 features: List[Feature],
                 actions: List[Action]) {
  //todo find by enum
  def findStat(statType: String): Option[Stat] =
    stats.find(s => s.statType.equals(statType.toLowerCase))
  //todo find by enum
  def findAbility(abilityType: String): Option[Ability] =
    abilityScores.find(a => a.abilityType.equals(abilityType.toLowerCase))

}
object Block{
  implicit val rw: RW[Block] = macroRW
}

@JSExportTopLevel("BlockAdapter")
class BlockAdapter() {
  @JSExport
  var block = Block("", Nil, Nil, Nil, Nil)

  @JSExport
  def setName(name: String): Unit =
    block = block.copy(name = name)

  @JSExport
  def putAbility(typeName: String, value: Int, modifier: Int): Unit =
    block = block.copy(abilityScores = Ability(typeName, value, modifier) :: block.abilityScores)

  @JSExport
  def putStat(typeName: String, statValue: StatValue): Unit =
    block = block.copy(stats = Stat(typeName, statValue) :: block.stats)

  @JSExport
  def putFeature(name: String, description: String): Unit =
    block = block.copy(features = Feature(name, description) :: block.features)

  @JSExport
  def putAction(action: Action): Unit =
    block = block.copy(actions = action :: block.actions)

  @JSExport
  def getBlock: String =
    write(block)
}